### PR TITLE
Expose findDanglingImpls to Python

### LINKED ISSUE #59885: Expose findDanglingImpls to Python
https://github.com/pytorch/pytorch/pull/46484/ is cool but it is not accessible from Python, which means you can't easily use it to test if backend extensions have dangling impls. A way to call it from Python would help a bit. Put the binding in torch/csrc/utils/python_dispatch.cpp Ideally it would be easy to get the (1) name of dangling impls and (2) filename/lineno where the dangling impl occurred.

### PR DESCRIPTION
**Pull request by alanwaketan**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#60827**

This diff exposed Dispatcher.findDanglingImpls to Python as _C._dispatch_find_dangling_impls.

Differential Revision: [D29416330](https://our.internmc.facebook.com/intern/diff/D29416330/)

### CODE REVIEW

File: torch/csrc/utils/python_dispatch.cpp [hunk 1]
```diff
@@ -184,6 +185,10 @@ void initDispatchBindings(PyObject* module) {
   m.def("_dispatch_check_all_invariants", []() {
     c10::Dispatcher::singleton().checkInvariants();
   });
+
+  m.def("_dispatch_find_dangling_impls", []() -> std::vector<c10::OperatorHandle> {
+    return c10::Dispatcher::singleton().findDanglingImpls();
```
Comment (alanwaketan) [inline torch/csrc/utils/python_dispatch.cpp:190]: This is the initial attempt. There are two things I'm not quite sure about:
1. If OperatorHandle has corresponding Python bindings or not. If not, should I covert all/some of the members into a dict?
2. How can I test the situation when there is a dangling implementation?

**Comment by ezyang** [comment]
don't forget to tag reviewers

**Comment by ezyang** [comment]
You should test this by manually triggering a dangling impl (e.g., add an extra `m.impl` call that doesn't have an `m.def` call anywhere) and check and see what the output looks like. Paste it here when you have done so.

File: torch/csrc/utils/python_dispatch.cpp (hunk 1 above)
Comment (ezyang) [reply torch/csrc/utils/python_dispatch.cpp]: 1. I'm pretty sure OperatorHandle doesn't have Python bindings. Giving it Python bindings is out of scope for this task. For this particular task, just design an output format that is adequate for debugging problems when there are dangling impls (for example, I'd like to see the source code file/line where the dangling implementation occurred). You'll need to actually exercise this case to see if you've done a good job, see below.
2. For a dangling implementation test that is run persistently in the test suite, the easiest way is to do a C++ extension test. Look at the msnpu extension test in test/test_cpp_extensions_aot.py for how it works

**Comment by alanwaketan** [comment]
> You should test this by manually triggering a dangling impl (e.g., add an extra `m.impl` call that doesn't have an `m.def` call anywhere) and check and see what the output looks like. Paste it here when you have done so.

Here is a quick one that I randomly added to the codebase.
name: aten::copy_imag
schema: (none)
Named: registered at caffe2/aten/src/ATen/core/NamedRegistrations.cpp:11 :: (none) [ fallthrough boxed ]

File: torch/csrc/utils/python_dispatch.cpp (hunk 1 above)
Comment (alanwaketan) [reply torch/csrc/utils/python_dispatch.cpp]: I was trying to build and import an extension that introduces a dangling impl but fail. I suspect `buck test` somehow messed up with the test directory such that the extension file cannot be found.

Here is the new test:
<pre>
def test_find_dangling_impls_ext(self):
    module = torch.utils.cpp_extension.load(
        name="msnpu_extension",
        sources=[
            os.path.dirname(os.path.abspath(__file__)) + "/cpp_extensions/msnpu_extension.cpp",
        ],
        extra_cflags=["-g"],
        verbose=True,
    )
</pre>

Here is how I ran the test:
`[.../fbcode/caffe2] buck test @mode/dev //caffe2/test:others -- test_find_dangling_impls_ext`.

Here is the output:
<pre>
Test output: 
> Repro command : $(cat "/tmp/tpx-20210629-202349.176685/61ad9903-4aa6-43f8-8409-b2edfef48535/execution_command")
FileNotFoundError: [Errno 2] No such file or directory: '/data/users/jwtan/fbsource/fbcode/buck-out/dev/gen/caffe2/test/others#binary,link-tree/cpp_extensions/msnpu_extension.cpp'
  File "/usr/local/fbcode/platform009/lib/python3.8/unittest/case.py", line 60, in testPartExecutor
    yield
  File "/usr/local/fbcode/platform009/lib/python3.8/unittest/case.py", line 676, in run
    self._callTestMethod(testMethod)
  File "/usr/local/fbcode/platform009/lib/python3.8/unittest/case.py", line 633, in _callTestMethod
    method()
  File "/data/users/jwtan/fbsource/fbcode/buck-out/dev/gen/caffe2/test/others#binary,link-tree/test_dispatch.py", line 895, in test_find_dangling_impls_ext
    module = torch.utils.cpp_extension.load(
  File "/data/users/jwtan/fbsource/fbcode/buck-out/dev/gen/caffe2/test/others#binary,link-tree/torch/utils/cpp_extension.py", line 1080, in load
    return _jit_compile(
  File "/data/users/jwtan/fbsource/fbcode/buck-out/dev/gen/caffe2/test/others#binary,link-tree/torch/utils/cpp_extension.py", line 1263, in _jit_compile
    version = JIT_EXTENSION_VERSIONER.bump_version_if_changed(
  File "/data/users/jwtan/fbsource/fbcode/buck-out/dev/gen/caffe2/test/others#binary,link-tree/torch/utils/_cpp_extension_versioner.py", line 45, in bump_version_if_changed
    hash_value = hash_source_files(hash_value, source_files)
  File "/data/users/jwtan/fbsource/fbcode/buck-out/dev/gen/caffe2/test/others#binary,link-tree/torch/utils/_cpp_extension_versioner.py", line 15, in hash_source_files
    with open(filename) as file:
</pre>

Comment (ezyang) [reply torch/csrc/utils/python_dispatch.cpp]: I think the cpp extension tests are OSS only, they don't run in fbcode. @alanwaketan, I know this is a bootcamp task, but consider taking the opportunity to setup an OSS build of PyTorch, it's an important part of working on the team and will tell you a little bit more about our workflow.

**Review by ezyang — APPROVED** [review]

File: torch/csrc/utils/python_dispatch.cpp (hunk 1 above)
Comment (alanwaketan) [reply torch/csrc/utils/python_dispatch.cpp]: Sure, let me add a todo to add an extension test case once I have completed the OSS setup. @ezyang Thanks for approving this patch.

**Pull request closed without merging**